### PR TITLE
Generate Address Handler

### LINKED ISSUE #391: Implement `generateAddress` API handler
#### Motivation
As mentioned in #314, generation of new P2PKH addresses requires implementing a new `generateAddress` API handler.

This method will consume @kronolynx crypto library to generate new "final keys", calculate their addresses and return those to the client along the used key path.

This method does not allow choosing the entire key path but only the account index. It will check instead what's the index of the last final key and create the next one.

#### Tasks
_Someone tackling this issue shall:_
1. Read and understand our documentation on [key derivation][hdwallets], [address formatting][addresses] and [slip32].
2. Implement the `generateAddress` API handler. This will:
2.1. Parse `params` as `GenerateAddressParams`. Throw `WRONG_PARAMS` otherwise.
2.2. Check if there's an unlocked wallet. Throw `NO_UNLOCKED_WALLET` otherwise.
2.3. Check if the account index does exist. Throw `WRONG_ACCOUNT` otherwise.
2.4. Get the maximum final key index as `k` (`0` by default).
2.5. If `label` is not set or length is `0`, set it to `Payment request #${k+1}`.
2.6. Check that `amount`, if set, is a positive number (`>= 0`). Throw `NEGATIVE_AMOUNT` otherwise. _In the future, we'll also check for enough balance_.
2.7. Check that `expires`, if set, falls in the range `[Date.now(), Date.now() + 31_536_000_000]`. Throw `PAST_EXPIRATION_DATE` or `TOO_FAR_EXPIRATION_DATE` accordingly.
2.8. Create a new final key using keypath `m/3'/4919'/i'/0/k`, where `i` is the account index and `k` is the final key index.
2.9. Store the final key in app state. Throw `WALLET_UPDATE_FAILURE` otherwise.
2.10. Persist app state (write a copy of the wallet into wallet storage). Throw `WALLET_STORE_FAILURE` otherwise.
2.11. Compute final key hash, trim to 20 bytes and encode using [slip32]. Throw `ADDRESS_ENCODING_FAILURE` otherwise.
2.12. Return a `GenerateAddressSuccess` with the used KeyPath, address and creation date.
3. Add a backend route for the `generateAddress` method in [routes.ts].
4. Create pertinent tests for the handler.

#### Deliverables
_A potential PR closing this issue shall provide:_
- A `generateAddress` handler that takes `GenerateAddressParams` as parameter and returns a `GenerateAddressSuccess` object.
- A route for the `generateAddress` method in [routes.ts].
- A test battery for the `generateAddress` handler.

#### Type definitions
```ts
type GenerateAddressParams = {
  account: number // we can optionally refine this type to be a inside range [0, 4294967295],
  label?: string,
  amount?: number,
  expirationDate?: number
}

type GenerateAddressSuccess = {
  kind: "SUCCESS",
  keyPath: KeyPath,
  address: string,
  creationDate: number
}
```

[routes.ts]: https://github.com/witnet/sheikah/blob/master/app/main/api/routes.ts
[hdwallets]: https://github.com/aesedepece/WIPs/blob/wip-adansdpc-hdwallets/wip-adansdpc-hdwallets.md
[addresses]: https://github.com/aesedepece/WIPs/blob/wip-adansdpc-addresses/wip-adansdpc-addresses.md
[slip32]: https://github.com/witnet/ts-slip32

### PR DESCRIPTION
**Pull request by anler**
# PR Prelude

Thank you for contributing to sheikah! :)

**Please fill in this template (by putting an `x` inside
the brackets) _before_ filing your PR:**

- [X] I have read and understood [CODE_OF_CONDUCT][code] document.
- [X] I have read and understood [CONTRIBUTING][cont] document.
- [X] I have read and understood [STYLEGUIDE][style] document.
- [X] I have included tests for the changes in my PR. If not, I have included a
  rationale for why I haven't.
- [X] My code is formatted and is accepted by `yarn fmt`.
- [X] **I understand my PR may be closed if it becomes obvious I didn't
  actually perform all of these steps.**

# Why this change is necessary and useful

[Please explain **in detail** why the changes in this PR are needed.]

Api handler for generating addresses. Closes #391 

[code]: https://github.com/witnet/sheikah/blob/master/.github/CODE_OF_CONDUCT.md
[cont]: https://github.com/witnet/sheikah/blob/master/.github/CONTRIBUTING.md
[style]: https://github.com/witnet/sheikah/blob/master/docs/STYLEGUIDE.md

### CODE REVIEW

**Comment by aesedepece** [comment]
Be careful with @Tommytrg's commits that slipped into the master branch but were dropped right after. They belong to a future PR.

**Comment by aesedepece** [comment]
Hi @witnet/devs friends, are we all OK to merge this as is and then address any particular issue separately?

File: app/main/api/handlers/getWallet.ts [hunk 1]
```diff
@@ -123,9 +123,12 @@ async function searchWallet(
  * @param {data: JsonSerializable}
  * @returns {Promise<Wallet>}
  */
-async function parseWallet(data: JsonSerializable): Promise<Wallet> {
+async function parseWallet(data: JsonSerializable, system: SubSystems): Promise<Wallet> {
   try {
-    return asRuntimeType(data, Wallet)
+    const wallet = asRuntimeType(data, Wallet)
+    system.appStateManager.update({ wallet })
```
Comment (mmartinbar) [inline app/main/api/handlers/getWallet.ts]: I'm under the impression that now that we store the `wallet` in the [AppState](https://github.com/witnet/sheikah/pull/436/files#diff-033d58029e7e5f4e0726a62f0605b05c), the app state should also be updated with the wallet in the `encryptWallet`handler

Comment (mmartinbar) [reply app/main/api/handlers/getWallet.ts]: Just pointing out things here to address them somewhere, but I'm ok with merging and then taking them into account separately :ok_hand:

**Comment by aesedepece** [comment]
This is to be superseded by a subsequent PR by @mariocao that will include all @anler's commits and some additional edits.

File: app/main/api/handlers/generateAddress.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+  if (amount && amount < 0) {
+    throw t.generateAddressErrors.NEGATIVE_AMOUNT
+  }
+
+  if (expirationDate) {
+    if (expirationDate < now) {
+      throw t.generateAddressErrors.PAST_EXPIRATION_DATE
+    }
+    if (expirationDate > now + 31_536_000_000) {
```
[the remark this answers is not in the document]
Comment (mariocao) [reply app/main/api/handlers/generateAddress.ts]: I thought the maximum should be 2 years?

**Comment by mariocao** [comment]
Superseded by #446

**Pull request closed without merging**